### DISCUSSION TITLE
Failed to load API definition after hosting the app on AWS

### DISCUSSION THREAD

**Discussion by kashyapm94** [fastapi/fastapi#6830] [Questions] [answered]
Hey, I am new to Fastapi and am migrating from Flask to here.

I am trying to host my app on AWS Api gateway using Serverless & Gitlab CI/CD. I found that currently there is only [mangum](https://github.com/jordaneremieff/mangum) that serves the purpose and I am using it.

When the CI/CD is successfully run on Gitlab, I get the endpoint of the app. When I am trying to access the endpoint via the browser, I get the following error:

![Screenshot from 2021-02-10 18-01-10](https://user-images.githubusercontent.com/35066641/107543941-fd778880-6bc9-11eb-9e4c-3b8686db0528.png)

But when I call the same endpoint using curl, I get the response back

The tree structure:

```
├── app
│   ├── api
│   │   ├── __init__.py
│   │   └── v1
│   │       ├── gcc_info.py
│   │       ├── __init__.py
│   │       ├── mscons.py
│   │       └── tokens.py
│   ├── core
│   │   ├── config.py
│   │   ├── decode_verify_jwt.py
│   │   ├── decorator.py
│   │   └── __init__.py
│   ├── crud
│   │   └── __init__.py
│   ├── database
│   │   └── __init__.py
│   ├── __init__.py
│   ├── main.py
│   ├── models
│   │   └── __init__.py
│   ├── schemas
│   │   ├── __init__.py
│   │   ├── payload.py
│   │   └── token.py
│   ├── services
│   │   ├── frontend_token.py
│   │   ├── frontend_uploader.py
│   │   ├── generator.py
│   │   ├── __init__.py
│   │   ├── processor.py
│   │   └── request_code.py
│   └── tests
│       └── __init__.py
├── package.json
├── package-lock.json
├── Pipfile
├── Pipfile.lock
├── README.md
└── serverless.yml
```
`main.py` is the entry point into the app and has the following content:

```
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from mangum import Mangum

from app.api.v1 import mscons, tokens, gcc_info
from app.core.config import settings

app = FastAPI(
    title=settings.PROJECT_NAME,
    version=0.1
)

app.add_middleware(
    CORSMiddleware,
    allow_origins=settings.BACKEND_CORS_ORIGINS,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

app.include_router(tokens.router)
app.include_router(mscons.router)
app.include_router(gcc_info.router)

handler = Mangum(app)

```

and the `serverless.yml` file is:

```
service: intense

plugins:
  - serverless-python-requirements
  - serverless-deployment-bucket

provider:
  name: aws
  runtime: python3.7
  stage: ${opt:stage, 'dev'}
  region: ${env:AWS_DEFAULT_REGION}
  endpointType: REGIONAL
  lambdaHashingVersion: 20201221
  apiGateway:
    shouldStartNameWithService: true

package:
  exclude:
    - app/tests/**
    - .coveragerc
    - .gitlab-ci.yml
    - package.json
    - package-lock.json
    - Pipfile
    - Pipfile.lock
    - README.md

custom:

  pythonRequirements:
    dockerizePip: non-linux

functions:
  intense:
    handler: app.main.handler
    timeout: 10
    memorySize: 128
    events:
      - http: ANY /
      - http: 'ANY /{proxy+}'

```

Any idea what is the mistake I am doing and why can't I see the swagger documentation hosted online?

The `openapi.json` of the swagger documentation when using on my local system is:

> {"openapi":"3.0.2","info":{"title":"Intense","version":"0.1"},"paths":{"/token":{"post":{"tags":["token"],"summary":"Login For Access Token","operationId":"login_for_access_token_token_post","requestBody":{"content":{"application/x-www-form-urlencoded":{"schema":{"$ref":"#/components/schemas/Body_login_for_access_token_token_post"}}},"required":true},"responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{"$ref":"#/components/schemas/Token"}}}},"422":{"description":"Validation Error","content":{"application/json":{"schema":{"$ref":"#/components/schemas/HTTPValidationError"}}}}}}},"/mscons/":{"post":{"tags":["mscons"],"summary":"Generate Mscons","description":"Generates the MSCONS based on the payload\n\n- **req**: FastAPI request\n- **payload**: Payload for the MSCONS","operationId":"generate_mscons_mscons__post","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/Payload"}}},"required":true},"responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{}}}},"403":{"description":"Forbidden"},"422":{"description":"Validation Error","content":{"application/json":{"schema":{"$ref":"#/components/schemas/HTTPValidationError"}}}}},"security":[{"OAuth2PasswordBearer":[]}]}},"/gcc/":{"get":{"tags":["gcc_info"],"summary":"Get Info","description":"Gets the username and gcc password from cognito userpool\n\n- **req**: FastAPI request","operationId":"get_info_gcc__get","responses":{"200":{"description":"Successful Response","content":{"application/json":{"schema":{}}}},"403":{"description":"Forbidden"}},"security":[{"OAuth2PasswordBearer":[]}]}}},"components":{"schemas":{"Body_login_for_access_token_token_post":{"title":"Body_login_for_access_token_token_post","required":["username","password"],"type":"object","properties":{"grant_type":{"title":"Grant Type","pattern":"password","type":"string"},"username":{"title":"Username","type":"string"},"password":{"title":"Password","type":"string"},"scope":{"title":"Scope","type":"string","default":""},"client_id":{"title":"Client Id","type":"string"},"client_secret":{"title":"Client Secret","type":"string"}}},"HTTPValidationError":{"title":"HTTPValidationError","type":"object","properties":{"detail":{"title":"Detail","type":"array","items":{"$ref":"#/components/schemas/ValidationError"}}}},"Payload":{"title":"Payload","required":["sender_mp_id","sender_mp_type","receiver_mp_id","receiver_mp_type","contact_name","mail_address","meter_pt_id","reference","position_number_lin","medium_obis","channel_obis","measured_var_obis","measure_type_obis","rate_obis","qty_wert","values"],"type":"object","properties":{"sender_mp_id":{"title":"Sender Mp Id","type":"string"},"sender_mp_type":{"title":"Sender Mp Type","type":"string"},"receiver_mp_id":{"title":"Receiver Mp Id","type":"string"},"receiver_mp_type":{"title":"Receiver Mp Type","type":"string"},"contact_name":{"title":"Contact Name","type":"string"},"mail_address":{"title":"Mail Address","type":"string"},"meter_pt_id":{"title":"Meter Pt Id","type":"string"},"reference":{"title":"Reference","type":"string"},"position_number_lin":{"title":"Position Number Lin","type":"string"},"medium_obis":{"title":"Medium Obis","type":"string"},"channel_obis":{"title":"Channel Obis","type":"string"},"measured_var_obis":{"title":"Measured Var Obis","type":"string"},"measure_type_obis":{"title":"Measure Type Obis","type":"string"},"rate_obis":{"title":"Rate Obis","type":"string"},"qty_wert":{"title":"Qty Wert","type":"array","items":{"type":"string"}},"values":{"title":"Values","type":"array","items":{"type":"array","items":{"anyOf":[{"type":"string"},{"type":"number"}]}}}}},"Token":{"title":"Token","required":["access_token"],"type":"object","properties":{"access_token":{"title":"Access Token","type":"string"}}},"ValidationError":{"title":"ValidationError","required":["loc","msg","type"],"type":"object","properties":{"loc":{"title":"Location","type":"array","items":{"type":"string"}},"msg":{"title":"Message","type":"string"},"type":{"title":"Error Type","type":"string"}}}},"securitySchemes":{"OAuth2PasswordBearer":{"type":"oauth2","flows":{"password":{"scopes":{},"tokenUrl":"token"}}}}}}

**Comment by ycd** [comment, marked as answer]
@kashyapm94 your stage name is `dev` right?

Can you try adding `/dev/` as root path? 


```python
app = FastAPI(
    title=settings.PROJECT_NAME,
    version=0.1,
    root_path="/dev/"
)

**Reply by ammarsaf** [reply]
Do we need to update `docs_url` and `openapi_url` too by adding `/dev/` in the path so we can see the docs? @ycd

**Comment by kashyapm94** [comment]
@ycd  That works. Thank you so much. Could you please explain to me how adding `root_path` solved the problem?

**Comment by ycd** [comment]
Sure, your app is running on some kind of port right for example `0.0.0.0:8080`, when you deploy your API and when you send a request to your `/dummy` endpoint. The request will go to `IP:8080/dummy`. But we have a also a prefix. so actually we have our API on `IP:8080/dev/dummy`. 

Adding a `root_path` adds a prefix to all your endpoints.

**Comment by kashyapm94** [comment]
Hey @ycd, I came back to the project today and realized one thing: if I use `root_path="/dev/"` in the local system when opening http://127.0.0.1:8000/docs, it shows the following error:
![Screenshot from 2021-03-01 15-51-24](https://user-images.githubusercontent.com/35066641/109514138-fe535a00-7aa5-11eb-84d2-bfb939002bf9.png)

I tried doing: http://127.0.0.1:8000/dev/docs, and it gives: `{"detail":"Not Found"}` error.

Any idea?

**Reply by ammarsaf** [reply]
Yes, it so weird that no one ever reply this. The problem after added `root_path` is, when we open docs, it shows this problem. How exactly to solve this? @tiangolo

**Reply by babaMar** [reply]
Try this: "Only specifying root_path in the FastAPI constructor equal to the API stage name worked for me"

**Reply by ammarsaf** [reply]
Can you show the code? @babaMar

**Reply by babaMar** [reply]
`app = FastAPI(title="My API title...", root_path=f'/{STAGE}')`, here STAGE is related to how you deploy the API (in my case  I'm deploying on AWS behind a gateway, so that would be `test`, `dev`, ... just after the API URL, before versioning

**Reply by ammarsaf** [reply]
I see. So the customized `root_path` is basically for the cloud deployment part and not for the local one (hence require logic), correct me if I'm wrong?

**Comment by agawronski** [comment]
This post has better SEO than the docs so leaving this here for others:
https://fastapi.tiangolo.com/advanced/behind-a-proxy/

**Comment by nyck33** [comment]
Edit: solved: https://www.reddit.com/r/FastAPI/comments/pznaoj/failed_to_load_api_definition_deployed_on_aws/?utm_source=share&utm_medium=web2x&context=3
I tried the suggested solution and still get the same error:
```
app = fastapi.FastAPI(
    title="coupon-backend",
    version=2.0,
    root_path="/beta/")
```
My url: https://oycvwin6xf.execute-api.us-west-2.amazonaws.com/beta/docs
https://oycvwin6xf.execute-api.us-west-2.amazonaws.com/beta/get-all-coupons works as expected.

**Comment by qangdatacube** [comment]
Hey @nyck33 . Try to increase your lambda function timeout. It solved my problem by increasing it from 3 seconds (by default) to 15 seconds. 
Check out this one: https://bobbyhadz.com/blog/aws-lambda-task-timed-out-after-seconds

**Comment by Rajesh2015** [comment]
> Hey @nyck33 . Try to increase your lambda function timeout. It solved my problem by increasing it from 3 seconds (by default) to 15 seconds. Check out this one: https://bobbyhadz.com/blog/aws-lambda-task-timed-out-after-seconds

Did it work for you with api versioning enabled

**Comment by Rajesh2015** [comment]
> Edit: solved: https://www.reddit.com/r/FastAPI/comments/pznaoj/failed_to_load_api_definition_deployed_on_aws/?utm_source=share&utm_medium=web2x&context=3 I tried the suggested solution and still get the same error:
> 
> ```
> app = fastapi.FastAPI(
>     title="coupon-backend",
>     version=2.0,
>     root_path="/beta/")
> ```
> 
> My url: https://oycvwin6xf.execute-api.us-west-2.amazonaws.com/beta/docs https://oycvwin6xf.execute-api.us-west-2.amazonaws.com/beta/get-all-coupons works as expected.

Did you manage to solve this with API versioning enabled .The solution with root path seems to be working without api versioning .Root_path ignored when we use fastapi_versioning

**Comment by babaMar** [comment]
Only specifying `root_path` in the FastAPI constructor equal to the API stage name worked for me

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue @kashyapm94  👍

**Comment by pymonger** [comment]
I ran into a similar issue and what worked for me was to define an environment variable that propagated the `stage` value in  `serverless.yml`:
```
provider:
  stage: ${opt:stage}, "dev"}

functions:
  app:
    environment:
      STAGE: ${self:provider.stage}
```
then in the `main.py` when instantiating the FastAPI object to dynamically set `root_path` if the environment variable is detected:
```
import os

app = FastAPI(
        title="Unity On-Demand REST API",
        version="0.0.1",
        description="Unity On-Demand Operations",
        root_path=f"/{os.environ.get('STAGE')}/" if "STAGE" in os.environ else None
)
```

Now I can get to the docs in both local development mode and in AWS API gateway/Lambda.

Maybe there's a cleaner way.

**Comment by Odzen** [comment]
Same here as @pymonger using serverless framework to deploy to AWS lambda:

```
import os

from fastapi import FastAPI
from mangum import Mangum

# Setting up path that is used to server the app
# To make it work live on AWS Lambda and locally

stage = os.environ.get('STAGE', None)
prefix = f"/{stage}" if stage else "/"

print("STAGE: ", stage)

app = FastAPI(title="Lambda", root_path=prefix)

handler = Mangum(app)
```

In this, it should work running locally, and in any other stage deployed to AWS Lambda with API Gateway.

**Reply by VpkPrasanna** [reply]
Tried the same method facing the same issue , please help me to resolve this issue 
Error Message 
![image](https://github.com/user-attachments/assets/c04528d7-7966-4b2b-927e-e39c113d7614)

Logs 
![image](https://github.com/user-attachments/assets/c75675c8-9c57-4027-a3fb-3c0946345c6d)

**Comment by tobiasfeil** [comment]
I bootstrapped my app with SAM, and it called the stage `Stage` in the beginning (note the uppercase `S`). I've always kept this name for the stage, but when trying 
```python
app = FastAPI(title='...', root_path='/Stage')
```
for some reason it didn't act as expected; only locally but not on APIGW. 

After changing it to lowercase `stage` in the Cloudformation Template and in the Python code, everything worked fine.

**Comment by ATNC** [comment]
Try to add 
`app = FastAPI(title='...', openapi_prefix='/stage/')`
and add openapi.json resource to API gateway

**Reply by ammarsaf** [reply]
`openapi_prefix` is deprecated

**Comment by PyDataBlog** [comment]
> `openapi_prefix` is deprecated

seems I have the same issue as you. Did you find a fix?

**Comment by hk2281** [comment]
итак, у меня такая же проблема, я запускаю приложение fast api в контейнере docker, и все с кодовым пространством хоста
после того, как я добавлю этот nginx.conf, все продолжает работать правильно```events {}
```http {
    server {
        listen 80;

        location /api/ {
            proxy_pass http://fastapi-app:8000/;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }


        location /docs {
            proxy_pass http://fastapi-app:8000/docs;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }


        location /openapi.json {
            proxy_pass http://fastapi-app:8000/openapi.json;
            proxy_set_header Host $host;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }

        location /images/ {
            alias /app/images/;
        }
    }
}```
